File: api/app/core/email/transport.py
"""Email transports. The transport is the network boundary — tests use ``InMemoryTransport``;
production uses ``SmtpTransport`` (aiosmtplib → MailHog locally, or a self-hosted SMTP)."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol, runtime_checkable


@dataclass
class SentMessage:
    to: str
    subject: str
    html: str


@runtime_checkable
class Transport(Protocol):
    async def send(self, *, to: str, subject: str, html: str, sender: str) -> None: ...


@dataclass
class InMemoryTransport:
    """Captures sent messages instead of dispatching — for tests and dry runs."""

    sent: list[SentMessage] = field(default_factory=list)

    async def send(self, *, to: str, subject: str, html: str, sender: str) -> None:
        self.sent.append(SentMessage(to=to, subject=subject, html=html))


@dataclass
class SmtpTransport:
    """Sends via SMTP using aiosmtplib (imported lazily so the dependency is only needed
    when actually dispatching)."""

    host: str = "127.0.0.1"
    port: int = 1025  # MailHog default
    username: str | None = None
    password: str | None = None
    use_tls: bool = False

    async def send(self, *, to: str, subject: str, html: str, sender: str) -> None:
        from email.message import EmailMessage

        import aiosmtplib

        msg = EmailMessage()
        msg["From"] = sender
        msg["To"] = to
        msg["Subject"] = subject
        msg.set_content("This message requires an HTML-capable client.")
        msg.add_alternative(html, subtype="html")

        await aiosmtplib.send(
            msg,
            hostname=self.host,
            port=self.port,
            username=self.username,
            password=self.password,
            use_tls=self.use_tls,
        )


def smtp_from_settings(settings: object) -> SmtpTransport:
    """Build an SmtpTransport from app settings (auth + TLS for a production relay)."""
    return SmtpTransport(
        host=settings.smtp_host,  # type: ignore[attr-defined]
        port=settings.smtp_port,  # type: ignore[attr-defined]
        username=settings.smtp_username or None,  # type: ignore[attr-defined]
        password=settings.smtp_password or None,  # type: ignore[attr-defined]
        use_tls=settings.smtp_use_tls,  # type: ignore[attr-defined]
    )

File: api-nest/src/email/transport.ts
/**
 * Email transports. Faithful port of api/app/core/email/transport.py. The transport is the
 * network boundary — tests/dev use InMemoryTransport; production uses SmtpTransport (nodemailer,
 * imported lazily so the dependency is only needed when actually dispatching).
 */

export interface SentMessage {
  to: string;
  subject: string;
  html: string;
}

export interface Transport {
  send(args: { to: string; subject: string; html: string; sender: string }): Promise<void>;
}

/** Captures sent messages instead of dispatching — for tests and dry runs. */
export class InMemoryTransport implements Transport {
  readonly sent: SentMessage[] = [];
  async send(args: { to: string; subject: string; html: string; sender: string }): Promise<void> {
    this.sent.push({ to: args.to, subject: args.subject, html: args.html });
  }
}

export interface SmtpConfig {
  host: string;
  port: number;
  username?: string | null;
  password?: string | null;
  useTls?: boolean;
}

/** Sends via SMTP using nodemailer (lazy import). */
export class SmtpTransport implements Transport {
  constructor(private readonly cfg: SmtpConfig) {}

  async send(args: { to: string; subject: string; html: string; sender: string }): Promise<void> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const nodemailer = require('nodemailer');
    const transporter = nodemailer.createTransport({
      host: this.cfg.host,
      port: this.cfg.port,
      secure: this.cfg.useTls ?? false,
      auth: this.cfg.username ? { user: this.cfg.username, pass: this.cfg.password ?? '' } : undefined,
    });
    await transporter.sendMail({
      from: args.sender,
      to: args.to,
      subject: args.subject,
      text: 'This message requires an HTML-capable client.',
      html: args.html,
    });
  }
}

/** Build a transport from env. Defaults to InMemory unless MAISHA_SMTP_ENABLED=true. */
export function buildTransport(): Transport {
  if (process.env.MAISHA_SMTP_ENABLED !== 'true') return new InMemoryTransport();
  return new SmtpTransport({
    host: process.env.MAISHA_SMTP_HOST ?? '127.0.0.1',
    port: Number(process.env.MAISHA_SMTP_PORT ?? 1025),
    username: process.env.MAISHA_SMTP_USERNAME || null,
    password: process.env.MAISHA_SMTP_PASSWORD || null,
    useTls: process.env.MAISHA_SMTP_USE_TLS === 'true',
  });
}
